### PR TITLE
Updated to GridMaze

### PR DESCRIPTION
**Pull request by mhyttsten**
- Added GridMaze as a target in Package.swift
- Made some code changes to core GridMaze environment
- Added full README.md explaining the environment

### CODE REVIEW

File: swift/Sources/OpenSpiel/GridMaze.swift [hunk 1]
```diff
@@ -396,43 +403,43 @@ public struct GridCell {
 public func SPACE(reward: Double = -1) -> GridCell {
   return GridCell(
     oneWordDescription: "SPACE", reward: reward,
-    isInitial: false, isTerminal: false,
+    isInitial: false, isTerminal: false, isGoal: false,
```
Comment (dan-zheng) [inline swift/Sources/OpenSpiel/GridMaze.swift]: These `GridCell` factories are more idiomatically written as static methods/properties:
```swift
extension GridCell {
  /// Creates a start cell with the given reward.
  static func start(reward: Double = -1) -> GridCell { fatalError() }
  /// Creates a goal cell with the given reward.
  static func goal(reward: Double) -> GridCell { fatalError() }
  /// Creates a hole cell with the given reward.
  static func hole(reward: Double) -> GridCell { fatalError() }
  /// Creates a wall cell with the given reward.
  static func wall(reward: Double) -> GridCell { fatalError() }
  /// A bounce cell.
  static var bounce: GridCell { fatalError() }
}
```

File: swift/Sources/OpenSpiel/GridMaze.swift [hunk 2]
```diff
@@ -100,16 +101,17 @@ public class GridMaze: GameProtocol {
...
-    cellTopSide: GridCell = BOUNCE(reward: -1.0),
-    cellBottomSide: GridCell = BOUNCE(reward: -1.0),
-    cellRightSide: GridCell = BOUNCE(reward: -1.0),
-    cellTopLeft: GridCell = BOUNCE(reward: -1.0),
-    cellTopRight: GridCell = BOUNCE(reward: -1.0),
-    cellBottomLeft: GridCell = BOUNCE(reward: -1.0),
-    cellBottomRight: GridCell = BOUNCE(reward: -1.0),
-    cellAllOthers: GridCell = SPACE(reward: -1.0)
+    rowCount: Int,
+    colCount: Int,
```
[the remark this answers is not in the document]
Comment (mhyttsten) [reply swift/Sources/OpenSpiel/GridMaze.swift]: Resolved in upcoming commit.

File: swift/Sources/OpenSpiel/GridMaze.swift [hunk 3]
```diff
@@ -100,16 +101,17 @@ public class GridMaze: GameProtocol {
...
+    cellsLeftSide: GridCell = WALL(),
```
[the remark this answers is not in the document]
Comment (mhyttsten) [reply swift/Sources/OpenSpiel/GridMaze.swift]: Great suggestion. Made this change in upcoming commit.

File: swift/Examples/GridMaze/main.swift [hunk 4]
```diff
@@ -17,154 +17,56 @@
...
-// Create policy table from qtable, a dictionary of type [String: [GridMaze.Action]]
-// For each informationState, this is a list of preferred actions
-// The value list only has >1 elements if all those actions have equivalent (and max) qvalues
-let policyTable = qtable.reduce([String: [GridMaze.Action]]()) { (dict, entry) in
-  let max = entry.value.max(by: { $0.qvalue < $1.qvalue })!.qvalue
-  let preferredActions = entry.value.reduce([GridMaze.Action]()) { list, entry in
-    var list = list
-    if entry.qvalue == max {
-      list.append(entry.action)
+    gameState.apply(actionToTake)
```
[the remark this answers is not in the document]
Comment (mhyttsten) [reply swift/Examples/GridMaze/main.swift]: Method apply is inherited from Spiel.StateProtocol though, or am I missing something? Keeping apply in upcoming commit

[the remark this answers is not in the document]
Comment (dan-zheng) [reply swift/Examples/GridMaze/main.swift]: I see! Keeping `apply` sounds good.

File: swift/Examples/GridMaze/README.md [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+* Cell [4,2] is a bounce cell. This means attempting to enter it from [3,2], [4,1], or [4,3] results in robot bouncing back to those cells. Cost will be -2 from [3,2] and -1 from [4,1] and [4,3].
+
+Now what about the white arrows with 50%? This indicates that cell [2,4] has random behavior.
+If the robot attempts to enter cell [2,4], it will with 50% probability end up in that cell, but it might also well end up in cell [3,4] with 50% probability.
+
+That's it for an overview of the environment, let's look at how interact with the environment through code!
+
+## Creating a GridMaze Environment
+This is how you create the maze environment in the picture above
+```
```
[the remark this answers is not in the document]
Comment (mhyttsten) [reply swift/Examples/GridMaze/README.md]: Great idea, didn't know about this. Updated in the next commit.

File: swift/Examples/GridMaze/README.md [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+This environment enables you to create and learn to solve grid mazes.
+Solving mazes is a nice problem to start with when learning the basics of reinforcement learning. 
+
+## Introduction
+The problem description of the GridMaze environment is that you have a Robot located on a starting cell in a grid maze world. 
+By moving Up, Down, Left, or Right, the robot should figure out the best strategy to move from the start cell to the goal cell of the maze.
+The robot gets a reward (positive value) or a cost (negative value) when moving to a new cell. The purpose of the game is for the robot to figure out how to move from the start to goal position and getting as much reward (or as little cost) as possible.
+
+Each cell in the grid world have different functions:
+* **SPACE**: The robot is free to enter this cell, but there is a at a certain cost for doing so (e.g. -1).
```
[the remark this answers is not in the document]
Comment (mhyttsten) [reply swift/Examples/GridMaze/README.md]: Agreed. Updated in the next commit.

[the remark this answers is not in the document]
Comment (mhyttsten) [reply swift/Examples/GridMaze/README.md]: Corrected in next commit.

File: swift/Examples/GridMaze/README.md [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+* **WALL**: This cell cannot be entered.
+* **BOUNCE**: The robot can attempt to enter this cell, but doing so will make it bounce back to the cell it came from, and get the associated reward/cost of that cell.
+* **HOLE**: The robot enters this cell at a typically big cost, and the game is over. In other words, it fails to reach the GOAL cell.
+
+Let's look at an example:
+![ExampleImage](https://github.com/deepmind/open_spiel/blob/master/swift/Examples/GridMaze/Images/GridMazeExample.png?raw=true)
+
+Describing each location in the maze by [row,column]:
+* Robot starts out in cell [1,1]. From this cell it can only move right or down, since a wall cell cannot be entered. Going right or down will cost -1.
+* The goal cell is at [4,4]. Entering this cell gives the robot 10 in reward, and the robot has successfully solve the maze (the T menas terminal cell).
```
[the remark this answers is not in the document]
Comment (mhyttsten) [reply swift/Examples/GridMaze/README.md]: Fixed in upcoming commit.

File: swift/Examples/GridMaze/README.md [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+* The space cell at [4,0] is a bit special. Since the cell at [4,5] is also space, it means the robot can go left in [4,0] and end up in [4,5]. Had [4,5] been a wall, going left from [4,0] would not have been a legal action.
+* Cell [4,2] is a bounce cell. This means attempting to enter it from [3,2], [4,1], or [4,3] results in robot bouncing back to those cells. Cost will be -2 from [3,2] and -1 from [4,1] and [4,3].
+
+Now what about the white arrows with 50%? This indicates that cell [2,4] has random behavior.
+If the robot attempts to enter cell [2,4], it will with 50% probability end up in that cell, but it might also well end up in cell [3,4] with 50% probability.
+
+That's it for an overview of the environment, let's look at how interact with the environment through code!
+
+## Creating a GridMaze Environment
+This is how you create the maze environment in the picture above
```
[the remark this answers is not in the document]
Comment (mhyttsten) [reply swift/Examples/GridMaze/README.md]: Fixed in an upcoming commit.

File: swift/Examples/GridMaze/README.md [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+* **GOAL**: The robot successfully solves the maze and typically gets a reward by walking to this cell. A maze can have more than one goal cell.
+* **WALL**: This cell cannot be entered.
+* **BOUNCE**: The robot can attempt to enter this cell, but doing so will make it bounce back to the cell it came from, and get the associated reward/cost of that cell.
+* **HOLE**: The robot enters this cell at a typically big cost, and the game is over. In other words, it fails to reach the GOAL cell.
+
+Let's look at an example:
+![ExampleImage](https://github.com/deepmind/open_spiel/blob/master/swift/Examples/GridMaze/Images/GridMazeExample.png?raw=true)
+
+Describing each location in the maze by [row,column]:
+* Robot starts out in cell [1,1]. From this cell it can only move right or down, since a wall cell cannot be entered. Going right or down will cost -1.
```
[the remark this answers is not in the document]
Comment (mhyttsten) [reply swift/Examples/GridMaze/README.md]: SGTM. Updated all cell locations to use backtick in an upcoming commit.

File: swift/Examples/GridMaze/README.md [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+mazeEnvExample[2, 3] = SPACE(reward: -2.0)
+mazeEnvExample[3, 2] = SPACE(reward: -2.0)
+mazeEnvExample[3, 4] = HOLE(reward: -100)
+mazeEnvExample[4, 0] = SPACE(reward: -1.0)
+mazeEnvExample[4, 2] = BOUNCE()
+mazeEnvExample[4, 4] = GOAL(reward: 1.0)
+mazeEnvExample[4, 5] = SPACE(reward: -1.0)
+mazeEnvExample[2, 4].entryJumpProbabilities = [(.Relative(1, 0), 0.5), (.Welcome, 0.5)]
+```
+As you can see, access to the cells in the environment is possible using the ```[row,column]``` syntax (GridMaze has overloaded the subscript operator).
```
[the remark this answers is not in the document]
Comment (mhyttsten) [reply swift/Examples/GridMaze/README.md]: Accepted suggestion. Changed all inline code segments to use single backticks.

File: swift/Examples/GridMaze/README.md [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+
+Now what about the white arrows with 50%? This indicates that cell [2,4] has random behavior.
+If the robot attempts to enter cell [2,4], it will with 50% probability end up in that cell, but it might also well end up in cell [3,4] with 50% probability.
+
+That's it for an overview of the environment, let's look at how interact with the environment through code!
+
+## Creating a GridMaze Environment
+This is how you create the maze environment in the picture above
+```
+var mazeEnvExample = GridMaze(rowCount: 6, colCount: 6)
```
[the remark this answers is not in the document]
Comment (mhyttsten) [reply swift/Examples/GridMaze/README.md]: Done.

File: swift/Sources/OpenSpiel/GridMaze.swift (hunk 1 above)
Comment (mhyttsten) [reply swift/Sources/OpenSpiel/GridMaze.swift]: SGTM. Changed code and doc in the upcoming commit.

**Comment by lanctot** [comment]
@dan-zheng: will you import it internally or shall I? If I do it, I need it to be the case that there are no more changes to be done. Would be cool to include this in Monday's update.

**Pull request merged**